Feature/rs php api

JudahDoupe opened the pull request:
- Added API calls for retrieving words and definitions for Review & Suggest
- Added API calls for up-voting and down-voting words through the comment system

JudahDoupe committed:
added a test API call

JudahDoupe committed:
- added rs_get_words API
 - uncommented lex_entry_read permissions

JudahDoupe committed:
added 'get_word_def' function witch extracts useful information from word entries.

JudahDoupe committed:
added api calls for getting the language of a project

JudahDoupe committed:
added an api call for getting comments for a word, and upvoting and downvoting words

JudahDoupe committed:
randomly selects words from DB in rs_get_words

irahopkinson commented on src/Api/Service/Sf.php:
```diff
@@ -369,6 +372,18 @@ public function project_getJoinRequests()
         return ProjectCommands::getJoinRequests($this->projectId);
     }
 
+    public function project_getLanguage()
```
This API method (and the one below it) is unnecessary. Use the **lex_projectDto** below and let the front end extract out what it needs itself. You'll quickly find you need lots of items from it anyway and it will only hit the server once rather than every time it needs a little bit of info.

Besides **ProjectCommands::readProject**, amongst other things, doesn't encode dates correctly. Which is why the Lexicon app uses the existing **lex_projectDto** method. Most of the time you should be able to use existing API methods rather than creating new ones.

irahopkinson commented on src/Api/Service/Sf.php:
```diff
@@ -657,7 +672,7 @@ public function lex_project_update($settings)
         return LexProjectCommands::updateProject($this->projectId, $this->userId, $settings);
     }
 
-    public function lex_baseViewDto()
```
This seems like a random change. Especially since you haven't change the corresponding name for every front end angular service that uses it. BTW DTO stands for Data Transport Object; it means we have collected together items (potentially from different sources) usually for a particular page so that we only hit the server once.

irahopkinson commented on src/Api/Service/Sf.php:
```diff
@@ -763,6 +778,23 @@ public function lex_comment_updateStatus($commentId, $status)
 
+    public function lex_comment_getByWord($wordId)
+    {
+        $project = new LexProjectModel($this->projectId);
```
If you look at most of the other methods in this API they are a one line call to another method (usually a command). This one belongs in **LexCommentCommands**. The reason for that is this file is particularly difficult to test well where as the thing it calls (the command) is easy to test. When you move it to **LexCommentCommands** make sure you have added tests for it in **LexCommentCommandsTest.php**.

This applies to all methods you have introduced below.

irahopkinson commented on src/Api/Service/Sf.php:
```diff
@@ -959,4 +991,59 @@ public function update_last_activity($newtime = null)
+
+    // -------------------------------- Review & Suggest Api ----------------------------------
+
+    public function rs_get_words($numWords = 64){
```
The namespace of **rs_** carries very little semantic content; **suggest_** would be better. It might be meaningful to you today because you have been using this abbreviation but not to others trying to understand your code. I realize we already have that problem with DTO (see my comment above) but we don't want to introduce new problems :).

irahopkinson reviewed (changes requested):
Your commit should not include changes to submodules. You might find `git submodule update --init --recursive` helps you sort this out.

JudahDoupe committed:
Moved all previous logic out of Sf.php to its respective commands file.

JudahDoupe commented:
I think I've made all the necessary changes.

 - I moved all of the logic out of the Sf.php file
 - I removed the unnecessary changes and functions
 - generally cleaned up the code
